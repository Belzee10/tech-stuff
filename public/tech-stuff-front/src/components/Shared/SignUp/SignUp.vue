<template>
  <v-container>
    <v-row>
      <v-col cols="4" offset="4">
        <v-form ref="form" v-model="valid" class="mt-10" lazy-validation>
          <v-text-field
            v-model="name"
            class="name"
            :rules="nameRules"
            label="Name"
            required
          ></v-text-field>
          <v-text-field
            v-model="lastName"
            class="last-name"
            :rules="lastNameRules"
            label="Last Name"
            required
          ></v-text-field>
          <v-text-field
            v-model="email"
            class="email"
            :rules="emailRules"
            label="E-mail"
            required
          ></v-text-field>
          <v-text-field
            v-model="password"
            class="password"
            :rules="passwordRules"
            type="password"
            label="Password"
            required
          ></v-text-field>
          <v-btn :disabled="!valid" color="primary" block @click="submit">
            Sign up
          </v-btn>
        </v-form>
        <v-alert v-if="errorRegister" class="mt-5" type="error">
          An error ocurred
        </v-alert>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'SignUp',
  data: () => ({
    valid: true,
    name: '',
    nameRules: [v => !!v || 'Name is required'],
    lastName: '',
    lastNameRules: [v => !!v || 'Last Name is required'],
    email: '',
    emailRules: [
      v => !!v || 'E-mail is required',
      v => /.+@.+\..+/.test(v) || 'E-mail must be valid'
    ],
    password: '',
    passwordRules: [v => !!v || 'Last Name is required'],
    error: false
  }),
  computed: {
    ...mapGetters(['errorRegister'])
  },
  methods: {
    ...mapActions(['register']),
    submit() {
      if (this.$refs.form.validate()) {
        this.register({
          name: this.name,
          lastName: this.lastName,
          email: this.email,
          password: this.password,
          role: 'member'
        }).then(() => this.$router.push({ name: 'home' }));
      }
    }
  }
};
</script>

<style lang="scss" scoped></style>
