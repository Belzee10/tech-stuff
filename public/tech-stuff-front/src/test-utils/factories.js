import { mount, createLocalVue } from '@vue/test-utils';
import mergeWith from 'lodash.mergewith';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

let vuetify;

beforeEach(() => {
  vuetify = new Vuetify();
});

const customizer = (_, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue;
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue;
  }
};

/**
 * create component wrapper
 * @param {*} Component
 * @param {Object} overrides
 */
const createWrapper = (Component, overrides) => {
  const defaultMountingOptions = {
    localVue,
    vuetify
  };
  return mount(
    Component,
    mergeWith(defaultMountingOptions, overrides, customizer)
  );
};

/**
 * create Vuex store
 * @param {Object} overrides
 */
const createStore = overrides => {
  const defaultStoreConfig = {};
  return new Vuex.Store(mergeWith(defaultStoreConfig, overrides, customizer));
};

export { createWrapper, createStore };
