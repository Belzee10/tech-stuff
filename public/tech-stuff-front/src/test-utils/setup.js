import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.config.productionTip = false;

Vue.use(Vuetify);

/**
 * silent "console.error"
 */
global.console = {
  error: jest.fn(),
  log: console.log,
  warn: console.warn,
  info: console.info,
  debug: console.debug
};
