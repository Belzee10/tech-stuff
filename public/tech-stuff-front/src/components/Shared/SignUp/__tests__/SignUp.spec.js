import { createWrapper, createStore } from '@/test-utils/factories';
import flushPromises from 'flush-promises';
import SignUp from '../SignUp.vue';

const data = {
  name: 'name',
  lastName: 'last name',
  email: 'email@email.com',
  password: '123',
  role: 'member'
};
const mocks = {
  $router: {
    push: jest.fn()
  }
};

describe('SignUp.vue', () => {
  test('should dispatch a "register" action with the correct payload and redirect to "home"', async () => {
    const store = createStore();
    store.dispatch = jest.fn(() => Promise.resolve());
    const wrapper = createWrapper(SignUp, {
      store,
      mocks
    });
    wrapper.find('.v-input.name input').setValue(data.name);
    wrapper.find('.v-input.last-name input').setValue(data.lastName);
    wrapper.find('.v-input.email input').setValue(data.email);
    wrapper.find('.v-input.password input').setValue(data.password);
    wrapper.find('.v-btn').trigger('click');
    expect(store.dispatch).toHaveBeenCalledWith('register', data);
    await flushPromises();
    expect(mocks.$router.push).toHaveBeenCalledWith({ name: 'home' });
  });

  test('should render an error Alert', async () => {
    const error = 'error';
    const store = createStore({
      getters: {
        errorRegister: () => error
      }
    });
    const wrapper = createWrapper(SignUp, {
      store,
      mocks
    });
    const alert = wrapper.find('.v-alert.error');
    expect(alert.exists()).toBeTruthy();
  });

  test('shoud render correctly', () => {
    const store = createStore();
    const wrapper = createWrapper(SignUp, {
      store,
      mocks
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
