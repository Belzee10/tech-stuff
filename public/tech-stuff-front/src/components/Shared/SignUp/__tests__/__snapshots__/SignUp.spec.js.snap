// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp.vue shoud render correctly 1`] = `
<div
  class="container"
>
  <div
    class="row"
  >
    <div
      class="col col-4 offset-4"
    >
      <form
        class="v-form mt-10"
        novalidate="novalidate"
      >
        <div
          class="v-input name theme--light v-text-field"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-46"
                  style="left: 0px; position: absolute;"
                >
                  Name
                </label>
                <input
                  id="input-46"
                  required="required"
                  type="text"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="v-input last-name theme--light v-text-field"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-49"
                  style="left: 0px; position: absolute;"
                >
                  Last Name
                </label>
                <input
                  id="input-49"
                  required="required"
                  type="text"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="v-input email theme--light v-text-field"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-52"
                  style="left: 0px; position: absolute;"
                >
                  E-mail
                </label>
                <input
                  id="input-52"
                  required="required"
                  type="text"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="v-input password theme--light v-text-field"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-55"
                  style="left: 0px; position: absolute;"
                >
                  Password
                </label>
                <input
                  id="input-55"
                  required="required"
                  type="password"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <button
          class="v-btn v-btn--block v-btn--contained theme--light v-size--default primary"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            
          Sign up
        
          </span>
        </button>
      </form>
       
      <!---->
    </div>
  </div>
</div>
`;
